### PR TITLE
Enable Order Creation in BO with a virtual product and a customer add…

### LINKED ISSUE #31594: Can't Create Order in BO when virtual product is in cart and customer address is from a Country that is not served by a carrier
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

In front office customers can complete orders for digital products even if their address country is not served by any carrier, it just requires to the country to be enabled.   When triying to do the same in the BO you are unable to create the order because it says "No carrier can be applied to this order" 

image

### Expected behavior

It is expected to have the same behavior as in the front office where you are able to create an order for a virtual product even if the customer country address  is not served by any carrier.

### Steps to reproduce

1. Create a Virtual product 
2. Remove the country of your choice from all the enabled carriers  in the store.
3. Go to Orders
4. Click Add new order
5. Select customer
6. select/add an address which uses the country you choose in step 2.
7. add a virtual product to the order
8. once you are here you wont be able to proceed to create the order

### PrestaShop version(s) where the bug happened

1.7.x -  8.1.x

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by cquintini**
…ress that is from a Country that is not served by any carrier.

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | In the front office customers can complete orders for digital products even if their address country is not served by any carrier, it just requires to the country to be enabled. When triying to do the same in the BO you are unable to create the order.
| Type?             | bug fix 
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     |  no
| How to test?      | Go to Orders  - Click Add new order - Select customer - select/add an address which uses the country THAT IS NOT SERVED BY ANY CARRIER.  - Add a virtual product to the order - you should be able to proceed with the order creation
| Fixed ticket?     | Fixes #31594

### CODE REVIEW

File: src/Adapter/Cart/QueryHandler/GetCartForOrderCreationHandler.php [hunk 1]
```diff
@@ -389,6 +390,9 @@ private function extractShippingFromLegacySummary(Cart $cart, array $legacySumma
     private function fetchCartDeliveryOptions(array $deliveryOptionsByAddress, int $deliveryAddressId)
     {
         $deliveryOptions = [];
+        if(empty($deliveryOptionsByAddress)){
```
Comment (kpodemski) [inline src/Adapter/Cart/QueryHandler/GetCartForOrderCreationHandler.php]:
```suggestion
        if (empty($deliveryOptionsByAddress)) {
```

**Review by kpodemski — COMMENTED** [review]
Hi @cquintini 

Thanks for your PR :) I think the change makes sense. I suggested one change to fix errors with coding standards.

File: src/Core/Domain/Cart/QueryResult/CartForOrderCreation/CartShipping.php [hunk 2]
```diff
@@ -66,6 +66,11 @@ class CartShipping
      */
     private $giftMessage;
 
+    /**
+     * @var bool
+     */
+    private $isVirtual;
```
Comment (kpodemski) [inline src/Core/Domain/Cart/QueryResult/CartForOrderCreation/CartShipping.php]:
```suggestion
    private $isVirtual;
```

File: src/Core/Domain/Cart/QueryResult/CartForOrderCreation/CartShipping.php [hunk 3]
```diff
@@ -148,4 +156,12 @@ public function getGiftMessage(): string
...
         return $this->giftMessage;
     }
+
+    /**
+     * @return bool
+     */
+    public function isVirtual(): bool
+    {
+        return $this->isVirtual;
+    }
```
Comment (kpodemski) [inline src/Core/Domain/Cart/QueryResult/CartForOrderCreation/CartShipping.php]:
```suggestion
    }
```

**Review by kpodemski — COMMENTED** [review]
coding standards

**Review by kpodemski — APPROVED** [review]

**Review by boherm — APPROVED** [review]
LGTM, thanks for your PR @cquintini!

**Review by aniszr — APPROVED** [review]
Hello @cquintini 

I tested your PR, the issue is solved.

See attached screenvideo:

https://github.com/PrestaShop/PrestaShop/assets/92626242/008be39b-791d-4e2e-b37c-c6c9b3b96825

Thanks!

**Comment by kpodemski** [comment]
thank you very much @cquintini ! Great contribution! 🎉

**Pull request merged**